packages/fx-core/src/component/deploy/azureDeployDriver.ts

```diff
@@ -0,0 +1,231 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import {
+  DeployStepArgs,
+  AzureResourceInfo,
+  DeployTypes,
+  DeployContext,
+  AzureUploadConfig,
+  AxiosOnlyStatusResult,
+} from "./deployTypes";
+import { checkMissingDeployArgs } from "../utils/common";
+import {
+  DeployExternalApiCallError,
+  DeployPrerequisiteError,
+  DeployTimeoutError,
+} from "../error/deployError";
+import { AzureAccountProvider, LogProvider } from "@microsoft/teamsfx-api";
+import { TokenCredentialsBase } from "@azure/ms-rest-nodeauth";
+import { BaseDeployDriver } from "./deployDriver";
+import { Base64 } from "js-base64";
+import * as appService from "@azure/arm-appservice";
+import { DeployConstant } from "../constant/deployConstant";
+import { default as axios } from "axios";
+import { AxiosZipDeployResult } from "./deployTypes";
+import { waitSeconds } from "../../common";
+
+export abstract class AzureDeployDriver extends BaseDeployDriver {
+  public static readonly AXIOS_INSTANCE = axios.create();
+
+  abstract pattern: RegExp;
+
+  async deploy(args: DeployTypes, context: DeployContext): Promise<void> {
+    const dist = checkMissingDeployArgs("deployDist", args.dist);
+    const src = checkMissingDeployArgs("deploySrc", args.src);
+
+    const resourceId = checkMissingDeployArgs("resourceId", args.resourceId);
+    const azureResource = this.parseResourceId(resourceId);
+    const azureCredential = await this.getAzureAccountCredential(context.azureAccountProvider);
+
+    return await this.azureDeploy(
+      { src: src, dist: dist, ignoreFile: args.ignoreFile },
+      azureResource,
+      azureCredential,
+      context
+    );
+  }
+
+  abstract azureDeploy(
+    args: DeployStepArgs,
+    azureResource: AzureResourceInfo,
+    azureCredential: TokenCredentialsBase,
+    context: DeployContext
+  ): Promise<void>;
+
+  /**
+   * check if resource id is legal and parse it
+   * @param resourceId deploy target
+   * @protected
+   */
+  protected parseResourceId(resourceId: string): AzureResourceInfo {
+    const result = resourceId.trim().match(this.pattern);
+    if (!result || result.length != 3) {
```

Comment:
It is better to make 3 as a constant of meaningful variable name instead of a magic number

Reply:
Although this 3 looks like a magic number, it actually corresponds to three backreferences in the regular expression